### PR TITLE
Support cpu and memory overcommit.

### LINKED ISSUE #44328: Overriding Total Cpu capacity. I.e., support cpu overcommit
Currently, it seems that kubelet get cpu and memory capacity programtically and there is no way to override the it. This is difficult when we need to customize the cpu and memory capacity to do some test.

### PR DESCRIPTION
**Pull request by haosdent**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Support oversold cpu and memory resources in node.

**Which issue(s) this PR fixes**:

Fixes #44328

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by dashpole** [comment]
/assign
I don't think we want to make this addition, but i'm curious why you want this feature.
/hold

File: pkg/kubelet/apis/config/types.go [hunk 1]
```diff
@@ -335,6 +335,11 @@ type KubeletConfiguration struct {
 	// This provide a "static" CPU list rather than the "dynamic" list by system-reserved and kube-reserved.
 	// This option overwrites CPUs provided by system-reserved and kube-reserved.
 	ReservedSystemCPUs string
+
+	// These options specifies the ratio to oversold node resources.
+	// If 0, these fields are ignored.
+	CPUOvercommitFactor uint
```
Comment (riking) [inline pkg/kubelet/apis/config/types.go]: If this is a ratio, why is it an integer? You appear to only be allowing 2x, 3x, 4x, etc resources advertised.

**Review by riking — CHANGES_REQUESTED** [review]

File: pkg/kubelet/apis/config/types.go (hunk 1 above)
Comment (haosdent) [reply pkg/kubelet/apis/config/types.go]: Thx riking. Ohoh, due to cpuset, we only support 2x, 3x, 4x. Or I change ratio to a better name?

Comment (riking) [reply pkg/kubelet/apis/config/types.go]: I would think that you use a different OCI interface than cpusets if you're oversubscribing..?

**Comment by zatricky** [comment]
With my SysAdmin hat I see this feature as being useful in an emergency in production. With my Developer hat, I see it as a feature far more likely to be misused and that better solutions already exist.

I recently had a cluster become CPU-exhausted as the hypervisor could not auto-scale the number of nodes or the resources available to the nodes. Being able to overcommit would have been helpful to avoid having to shut down each node in order to upgrade their CPU allocations until a later more convenient time when deployment autoscaling would have reduced the number of operational pods.

The scenario would have been better handled if we had better monitoring in place and a better hypervisor stack ; thus I say the benefit would have been limited and temporary.

**Comment by julienlau** [comment]
> /assign I don't think we want to make this addition, but i'm curious why you want this feature. /hold

Hi,

I think this issue is still relevant !

This would be very helpful to improve cpu utilisation on Kubernetes cluster used as compute plane for example for Spark.
As explained in this Spark ticket, with Spark there is an assumption made that when you request N cores for parallel compute, that N cores would be 100% used.
In this way, Spark automatically cpu limits and cpu requests equal to N.
This is very inefficient, since most workload also do IO and are using effectively not more than 50% cpu usage.

https://issues.apache.org/jira/browse/SPARK-32744

From my experience as consultant on numerous projects, I see a lot of Kubernetes cluster used as Spark backend having really low CPU  / memory real usage.

If we could set overcommit factors for memory and cpu on nodes spec that would be a very elegant and convenient way to improve resources usage.

**Pull request closed without merging**